core/trino-main/src/main/java/io/trino/sql/planner/optimizations/DeterminePartitionCount.java

```diff
@@ -308,6 +315,24 @@ private static double getSourceNodesOutputStats(PlanNode root, ToDoubleFunction<
                 .sum();
     }
 
+    private static boolean isEligibleRemoteExchangePresent(PlanNode root)
+    {
+        return PlanNodeSearcher.searchFrom(root)
+                .where(node -> node instanceof ExchangeNode exchangeNode && isEligibleRemoteExchange(exchangeNode))
+                .matches();
+    }
+
+    private static boolean isEligibleRemoteExchange(ExchangeNode exchangeNode)
+    {
+        if (exchangeNode.getScope() != REMOTE || exchangeNode.getType() != REPARTITION) {
+            return false;
+        }
+        PartitioningHandle partitioningHandle = exchangeNode.getPartitioningScheme().getPartitioning().getHandle();
+        return !partitioningHandle.isScaleWriters()
```

Comment:
nit: this is probably enforced by `PlanNodeSearcher.searchFrom(plan).whereIsInstanceOfAny(INSERT_NODES).matches();` check.

@gaurav8297 ?